KT-35811: Type parameter angle brackets followed by equal sign are parsed incorrectly if whitespace is missing

kralliv opened the pull request:
As described in [KT-35811](https://youtrack.jetbrains.com/issue/KT-35811) writing something like `val a: List<String>= listOf()`, where there is no whitespace in between the `>` and `=`, results in a parse error.

With the changes in this pull request, `GTEQ` will no longer be tokenized by the lexer directly, but instead will be joined to a complex token, like it is done with `SAFE_CALL` and `ELVIS`. This way, the `GT` and `EQ` can be retained as separat token types in the relevant sections, while remaining a `GTEQ` in all the other places.

Both `_JetLexer` and `ParsingTestGenerated` were generated by gradle.

kralliv commented:
How exact must the diagnostic error overlay with the original parse error? Is it simply enough, that a diagnostic error is reported roughly at the parse error's position or do they have to match in their ranges? The nature of the parse error causes the error to span multiple elements in the tree.

Also, what major version should I target with this fix and therefore the `LanguageFeature`?

dzharkov commented on compiler/frontend/src/org/jetbrains/kotlin/diagnostics/rendering/DefaultErrorMessages.java:
```diff
@@ -625,6 +625,7 @@ public static DiagnosticRenderer getRendererForDiagnostic(@NotNull UnboundDiagno
                                                "This call will resolve to another declaration: {0}. " +
                                                "See https://youtrack.jetbrains.com/issue/KT-49276 for more details. " +
                                                "Please specify a progression type of argument explicitly through explicit cast to resolve to a proper declaration", COMPACT);
+        MAP.put(EXPRESSION_AFTER_TYPE_REFERENCE_WITHOUT_SPACING_NOT_ALLOWED, "Types must have a whitespace after the '>' if followed by an expression body");
```
I would add a piece of promise that this would be allowed since 1.8

dzharkov reviewed (changes requested):
Looks nice!
But some tests would be appreciated
The simplest thing would be adding a couple of them to `compiler/testData/diagnostics/tests`: one with `// !LANGUAGE: -AllowExpressionAfterTypeReferenceWithoutSpacing` and another with `// !LANGUAGE: +AllowExpressionAfterTypeReferenceWithoutSpacing`

kralliv commented:
I added tests and generated the test classes. I also added support for FIR.

dzharkov commented:
Yeah, tests look good!
But I'd remove the checker for FIR since once we release it (hardly soon), there will be no need to support obsolete version like 1.7. At the same time, the checker even that simple would eat a piece of CPU time, while actually that's not necessary.
Sorry for not being told that beforehand :(

Also, please rebase the branch on master.

kralliv commented:
Alright, I guess should have asked about the FIR first, given that I wasn't sure about it. I removed the FIR support again, squashed the commits and rebased it on the latest version of the master branch.

dzharkov commented:
Thank you for your contribution! Merged manually.
